spring-integration-jms/src/test/java/org/springframework/integration/jms/JmsOutboundGatewayTests.java

```diff
@@ -157,6 +159,7 @@ public void testReplyContainerRecovery() throws Exception {
 		}
 		finally {
 			gateway.stop();
+			exec.shutdownNow();
```

Comment:
Oh! That's cool catch!

I guess we need to revise all the tests in project for similar vulnerability 😄 

I wonder how did you come here? What helped you to figure out the culprit?

Thanks

Reply:
I was just lucky; I happened to take a `jstack <pid>` and noticed hundreds of threads including many listener container threads; since we have `@DirtiesContext` on all tests, something didn't seem right. Hence I found the bug in the gateway. I also noticed a bunch of task executor threads so I started looking at some of the tests.